### PR TITLE
Distributed KNN

### PR DESCRIPTION
**Pull request by ethanglaser**
# Description
Initial working implementation of distributed brute force KNN algorithm. Further optimizations will be included in subsequent PRs.

### Changes include:
-Restructure existing gpu knn implementation with a general brute force [file ](https://github.com/ethanglaser/oneDAL/blob/dev/eglaser-distr-knn/cpp/oneapi/dal/algo/knn/backend/gpu/infer_kernel_brute_force_dpc.cpp) as well as [single ](https://github.com/ethanglaser/oneDAL/blob/dev/eglaser-distr-knn/cpp/oneapi/dal/algo/knn/backend/gpu/infer_kernel_impl_dpc.hpp) and [distributed](https://github.com/ethanglaser/oneDAL/blob/dev/eglaser-distr-knn/cpp/oneapi/dal/algo/knn/backend/gpu/infer_kernel_impl_dpc_distr.hpp) implementation files
-Distributed Algorithm:
&nbsp;&nbsp;&nbsp;&nbsp;-[Splitting](https://github.com/ethanglaser/oneDAL/blob/dev/eglaser-distr-knn/cpp/oneapi/dal/algo/knn/backend/gpu/infer_kernel_impl_dpc_distr.hpp#L542-L559) train data into a sequence of blocks to be iteratively evaluated on
&nbsp;&nbsp;&nbsp;&nbsp;-Main iterative [loop](https://github.com/ethanglaser/oneDAL/blob/dev/eglaser-distr-knn/cpp/oneapi/dal/algo/knn/backend/gpu/infer_kernel_impl_dpc_distr.hpp#L626) that calls primitive functionality and triggers callback operator
&nbsp;&nbsp;&nbsp;&nbsp;-KNN callback [operator](https://github.com/ethanglaser/oneDAL/blob/dev/eglaser-distr-knn/cpp/oneapi/dal/algo/knn/backend/gpu/infer_kernel_impl_dpc_distr.hpp#L219) that merges the results from the current training block with the best results from previous blocks
&nbsp;&nbsp;&nbsp;&nbsp;-Send receive replace MPI communicator [functionality ](https://github.com/ethanglaser/oneDAL/blob/dev/eglaser-distr-knn/cpp/oneapi/dal/algo/knn/backend/gpu/infer_kernel_impl_dpc_distr.hpp#L692-L700) that cycles blocks of training data across devices
-Minor restructuring and revision of search primitives
-Restructuring of test files similar to other distributed algos, with addition of [spmd tests](https://github.com/ethanglaser/oneDAL/blob/dev/eglaser-distr-knn/cpp/oneapi/dal/algo/knn/test/spmd.cpp)
-Addition of inference functions to oneDAL general test [fixtures](https://github.com/ethanglaser/oneDAL/blob/dev/eglaser-distr-knn/cpp/oneapi/dal/test/engine/fixtures.hpp)
-Creation of [mpi](https://github.com/ethanglaser/oneDAL/blob/dev/eglaser-distr-knn/samples/oneapi/dpc/mpi/sources/knn_bf_distr_mpi.cpp) and [ccl](https://github.com/ethanglaser/oneDAL/blob/dev/eglaser-distr-knn/samples/oneapi/dpc/ccl/sources/knn_bf_distr_ccl.cpp) samples

### CODE REVIEW

**Comment by samir-nasibli** [comment]
Please rebase your branch first

File: cpp/oneapi/dal/algo/knn/test/fixture.hpp [hunk 1]
```diff
@@ -0,0 +1,258 @@
+/*******************************************************************************
+* Copyright 2021 Intel Corporation
```
Comment (napetrov) [inline cpp/oneapi/dal/algo/knn/test/fixture.hpp:2]: new files should go with 2023

File: cpp/oneapi/dal/algo/knn/test/spmd.cpp [hunk 2]
```diff
@@ -0,0 +1,598 @@
...
+                         "[external-dataset][knn][integration][spmd][test]", \
+                         knn_reg_types)
+
+#define KNN_SPMD_REG_BF_EXTERNAL_TEST(name)                                  \
+    TEMPLATE_LIST_TEST_M(knn_spmd_test,                                      \
+                         name,                                               \
+                         "[external-dataset][knn][integration][spmd][test]", \
+                         knn_reg_bf_types)
+
+KNN_SPMD_SMALL_TEST("knn nearest points test predefined 7x5x2") {
```
Comment (KulikovNikita) [inline cpp/oneapi/dal/algo/knn/test/spmd.cpp:170]: mismatch in dimensions and name of the case

File: cpp/oneapi/dal/algo/knn/test/spmd.cpp [hunk 3]
```diff
@@ -0,0 +1,598 @@
...
+    this->exact_nearest_indices_check(x_train_table, x_infer_table, infer_result);
+}
+
+KNN_SPMD_CLS_SYNTHETIC_TEST("distributed knn nearest points test random uniform 513x301x17") {
+    SKIP_IF(this->get_policy().is_cpu());
+    SKIP_IF(this->not_available_on_device());
+    SKIP_IF(this->not_float64_friendly());
+    SKIP_IF(this->is_kd_tree);
+
+    this->set_rank_count(10);
```
Comment (KulikovNikita) [inline cpp/oneapi/dal/algo/knn/test/spmd.cpp:210]: Too many ranks) Please consider 2, 3, 4

File: cpp/oneapi/dal/algo/knn/test/spmd.cpp [hunk 4]
```diff
@@ -0,0 +1,598 @@
...
+    const table y_train_table = this->arange(train_row_count);
+
+    const auto knn_desc = this->get_descriptor(train_row_count, 1);
+
+    auto infer_result =
+        this->train_and_infer_spmd(knn_desc, x_train_table, y_train_table, x_infer_table);
+
+    this->exact_nearest_indices_check(x_train_table, x_infer_table, infer_result);
+}
+/*
```
Comment (KulikovNikita) [inline cpp/oneapi/dal/algo/knn/test/spmd.cpp:302]: What is the reason why this cases are commented

File: cpp/oneapi/dal/backend/primitives/search/search_dpc.cpp [hunk 5]
```diff
@@ -27,6 +27,14 @@
 
 namespace oneapi::dal::backend::primitives {
 
+// // TODO: figure out this function
+template <typename Float>
+std::int64_t get_block_size() {
+    constexpr std::int64_t result = 128 * 4096 * 8 / sizeof(Float);
```
Comment (KulikovNikita) [inline cpp/oneapi/dal/backend/primitives/search/search_dpc.cpp:33]:
```suggestion
    constexpr std::int64_t result = 128 * 4096 * 8 / sizeof(Float);
```
```suggestion
    constexpr std::int64_t result = 1024 * 4096 / sizeof(Float);
```

File: samples/oneapi/dpc/ccl/sources/knn_bf_distr_ccl.cpp [hunk 6]
```diff
@@ -0,0 +1,87 @@
...
+    auto rank_id = comm.get_rank();
+    auto rank_count = comm.get_rank_count();
+
+    auto x_train_vec = split_table_by_rows<float>(queue, x_train, rank_count);
+    auto y_train_vec = split_table_by_rows<float>(queue, y_train, rank_count);
+    auto x_test_vec = split_table_by_rows<float>(queue, x_test, rank_count);
+    auto y_test_vec = split_table_by_rows<float>(queue, y_test, rank_count);
+
+    // First value is number of classes, second is number of neighbors
+    // Voting mode and distance impl should resort to default
```
Comment (KulikovNikita) [inline samples/oneapi/dpc/ccl/sources/knn_bf_distr_ccl.cpp:55]: Why should it resort to defaults?

**Review by Alexandr-Solovev — APPROVED** [review]

**Pull request merged**